limitador/src/storage/mod.rs

```diff
@@ -21,38 +23,230 @@ pub enum Authorization<'c> {
     Limited(&'c Counter), // First counter found over the limits
 }
 
-pub trait Storage: Sync + Send {
-    fn get_namespaces(&self) -> Result<HashSet<Namespace>, StorageErr>;
-    fn add_limit(&self, limit: &Limit) -> Result<(), StorageErr>;
-    fn get_limits(&self, namespace: &Namespace) -> Result<HashSet<Limit>, StorageErr>;
-    fn delete_limit(&self, limit: &Limit) -> Result<(), StorageErr>;
-    fn delete_limits(&self, namespace: &Namespace) -> Result<(), StorageErr>;
+pub struct Storage {
+    limits: RwLock<HashMap<Namespace, HashSet<Limit>>>,
+    counters: Box<dyn CounterStorage>,
+}
+
+pub struct AsyncStorage {
+    limits: RwLock<HashMap<Namespace, HashSet<Limit>>>,
+    counters: Box<dyn AsyncCounterStorage>,
+}
+
+impl Storage {
+    pub fn new() -> Self {
+        Self {
+            limits: RwLock::new(HashMap::new()),
+            counters: Box::new(InMemoryStorage::default()),
+        }
+    }
+
+    pub fn with_counter_storage(counters: Box<dyn CounterStorage>) -> Self {
+        Self {
+            limits: RwLock::new(HashMap::new()),
+            counters,
+        }
+    }
+
+    pub fn get_namespaces(&self) -> HashSet<Namespace> {
+        self.limits.read().unwrap().keys().cloned().collect()
+    }
+
+    pub fn add_limit(&self, limit: Limit) {
+        let namespace = limit.namespace().clone();
+        self.limits
+            .write()
+            .unwrap()
+            .entry(namespace)
+            .or_default()
+            .insert(limit);
+    }
+
+    pub fn get_limits(&self, namespace: &Namespace) -> HashSet<Limit> {
+        match self.limits.read().unwrap().get(namespace) {
+            Some(limits) => limits.iter().cloned().collect(),
+            None => HashSet::new(),
+        }
+    }
+
+    pub fn delete_limit(&self, limit: &Limit) -> Result<(), StorageErr> {
+        let mut limits = HashSet::new();
+        limits.insert(limit.clone());
+        self.counters.delete_counters(limits)?;
+
+        let mut limits = self.limits.write().unwrap();
```

Comment:
I am always a little hesitant with so many unwrap but I understand likelyhood is low. Do you think we should manage results instead of panicing. Maybe I am being too conservative.

Reply:
Well… again, nothing new here. But these `.unwrap()` are on `LockResult<RwLockReadGuard>` which would return an `Err` if the lock is "poisoned". That happens when a thread panic'ed while holding an exclusive lock. This indicates that whatever this lock was guarding probably is "corrupted". But so that would mean: something went real bad in the past, for which we didn't cleaning recover from _and_ some other thread is now trying to use that (possibly) corrupted state. 
Until we find a strategy to recover from a poisoned state, I think panic'ing completely is probably the best option anyhow. One way would be to explicitly table flip and `process::exit(1)` on such a situation… unsure how much I want to litter the code with these conditions tho. 

Again, if we want to change that (which I'm certainly open to, if only to start the discussion on how to achieve that), I wouldn't include these functional changes in this PR.